pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -242,8 +242,15 @@ func (m *ManagerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.P
 	pod := attrs.Pod
 	// TODO: Reuse devices between init containers and regular containers.
 	for _, container := range pod.Spec.InitContainers {
-		if err := m.allocateContainerResources(pod, &container); err != nil {
-			return err
+		for resourceKey := range container.Resources.Limits {
+			resource := string(resourceKey)
+			_, registeredResource := m.allDevices[resource]
```

Comment:
Should also be there an additional check that this resource is also requested by regular container. Only then return error?

Reply:
because init-containers run sequentially, we did reuse between init containers in scheduler, that is, if we have two init containers, each request 1 device, then we expect 1 device in total in scheduler, but expect 2 in deviceplugin. The check would not cover the scenario if we loosen the check.

Reply:
This is indeed not optimal. Thinking about this, I am wondering whether we should try to implement option 2) and get it into 1.9.

Reply:
Would like to work on that, but we may still first need to reach an agreement on the way to initialize devices, which could essentially affect the way to implement it.
As @RenaudWasTaken  implied, if devices could not be reused without calling plugin to make re-allocation,  option 2) could not work; and as @vikaschoudhary16  commented, if we treat devices stateless in plugins, we could thoroughly generate setting for all of the containers that share a device at once in admission.And thus, the option could be the final solution to cover all the scenarios.

I think if we could not reach an agreement on the point, it's hard to decide how to make the change, and we may fail to see if we could benefit from it. If so, considering the fact that the feature is still in alpha, I would prefer we forbid the use of Device Plugin Resources in Init Containers for the 1.9 release, and investigate for a solution that could certainly cover user's practical cases.

Reply:
IMHO option 2) is basically a subset of a feature allowing to share Devices (and it's really bad because you don't control which containers will share these devices).
We don't support Device sharing therefore we should not try to implement option 2).

Denying Device Plugin Resources altogether seem to be the only short term solution.

Reply:
@RenaudWasTaken Calling it Device `Sharing` is a misnomer, it is rather device 're-use'/'initialisation'. That too between init containers and app containers and not among app containers.
With 2), still containers have exclusive access to devices.